apps/dcos_overlay/src/dcos_overlay_poller.erl

```diff
@@ -110,61 +101,141 @@ code_change(_OldVsn, State, _Extra) ->
 %%%===================================================================
 
 update_poll_period(OldPollPeriod) when OldPollPeriod*2 =< ?MAX_POLL_PERIOD ->
-    OldPollPeriod*2;
+    OldPollPeriod * 2;
 update_poll_period(_) ->
     ?MAX_POLL_PERIOD.
 
-parse_response(State0 = #state{known_overlays = KnownOverlays}, AgentInfo) ->
-    IP0 = maps:get(<<"ip">>, AgentInfo),
-    IP1 = process_ip(IP0),
-    State1 = State0#state{ip = IP1},
+poll(State0) ->
+    prometheus_counter:inc(overlay, polls_total, [], 1),
+    Begin = erlang:monotonic_time(),
+    PollResponse = dcos_net_mesos:poll("/overlay-agent/overlay"),
+    prometheus_histogram:observe(overlay, poll_duration_seconds, [],
+        erlang:monotonic_time() - Begin),
+    case PollResponse of
+        {error, Error} ->
+            prometheus_counter:inc(
+              overlay, poll_errors_total, [], 1),
+            lager:warning("Overlay Poller could not poll: ~p", [Error]),
+            State0#state{poll_period = ?MIN_POLL_PERIOD};
+        {ok, Data} ->
+            NewPollPeriod = update_poll_period(State0#state.poll_period),
+            case parse_response(State0, Data) of
+                {ok, State1} ->
+                    State1#state{poll_period = NewPollPeriod};
+                {error, Error} ->
+                    lager:error("Failed to parse overlay response due to ~p",
+                                [Error]),
+                    exit(Error)
+            end
+    end.
+
+parse_response(State0 = #state{netlink = Pid, known_overlays = KnownOverlays},
+               AgentInfo) ->
+    AgentIP0 = maps:get(<<"ip">>, AgentInfo),
+    AgentIP = process_ip(AgentIP0),
+    State1 = State0#state{ip = AgentIP},
     Overlays = maps:get(<<"overlays">>, AgentInfo, []),
     NewOverlays = Overlays -- KnownOverlays,
-    lists:foldl(fun add_overlay/2, State1, NewOverlays).
+    Results = [catch add_overlay(Pid, Overlay, AgentIP)
+               || Overlay <- NewOverlays],
+    {Successes, Failures} =
+        lists:partition(fun(ok) -> true; (_) -> false end, Results),
+    prometheus_gauge:set(overlay, networks_configured, [],
+        length(Successes) + length(KnownOverlays)),
```

Comment:
Eventually, this metrics will be equal to `length(Overlays)` anyway cuz `known_overlays` contains all overlay networks. Which is fine, but code could be easier in that case.

Reply:
If `eventually` is fine here, I am happy to adjust.